### PR TITLE
Enable subfield pruning to pass through arbitrary() function

### PR DESCRIPTION
**Pull request by gparai**
Subfield pruning applies to queries accessing columns of complex types: maps, arrays and structs.
Queries that use only some indices, keys or fields of these columns are optimized so that only
necessary data is extracted during the table scan. This PR enables subfield pruning to pass
through the `arbitrary()` aggregate function. 

For example, in the following query only subfield `a` will be extracted from struct `c` during table scan.

```
CREATE TABLE t(id bigint, c row(a bigint, b varchar));

SELECT id, arbitrary(c).a 
FROM t 
GROUP BY 1;
``` 

== NO RELEASE NOTE ==

### CODE REVIEW

**Comment by gparai** [comment]
@mbasmanova can you please take a look at the PR and let me know your thoughts. The general idea is to pass along the subfield information from the project preceding the aggregate in the context and then use the context in the aggregate to generate the pass-through fields for the scan.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java [hunk 1]
```diff
@@ -130,8 +138,25 @@ public PlanNode visitAggregation(AggregationNode node, RewriteContext<Context> c
...
-            for (AggregationNode.Aggregation aggregation : node.getAggregations().values()) {
-                aggregation.getArguments().forEach(expression -> subfieldExtractor.process(castToExpression(expression), context.get()));
+            for (VariableReferenceExpression variable : node.getAggregations().keySet()) {
+                AggregationNode.Aggregation aggregation = node.getAggregations().get(variable);
+                // Aggregate calls with pass-through functions should allow sub-field pruning
+                if (PASS_THROUGH_FUNCTIONS.contains(metadata.getFunctionManager().getFunctionMetadata(aggregation.getCall().getFunctionHandle()).getName())) {
+                    aggregation.getArguments().forEach(expression ->
+                    {
+                        Expression var = new SymbolReference(variable.getName());
+                        if (context.get().getPassThroughFields(var).isPresent()) {
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java]: I'm not sure I understand this logic. Currently, there is only one pass-through aggregate function and it has only one argument. Hence, it is difficult for me to think though how would a multi-argument pass through function would work here. I suggest to simplify this change to work just for the `arbitrary` aggregate. I tried the following and the tests pass.

```
private static final QualifiedFunctionName ARBITRARY_AGGREGATE_FUNCTION = QualifiedFunctionName.of(DEFAULT_NAMESPACE, "arbitrary");

            for (Map.Entry<VariableReferenceExpression, AggregationNode.Aggregation> entry : node.getAggregations().entrySet()) {
                VariableReferenceExpression variable = entry.getKey();
                AggregationNode.Aggregation aggregation = entry.getValue();

                // Aggregate calls with pass-through functions should allow sub-field pruning
                QualifiedFunctionName aggregateName = metadata.getFunctionManager().getFunctionMetadata(aggregation.getCall().getFunctionHandle()).getName();
                if (ARBITRARY_AGGREGATE_FUNCTION.equals(aggregateName)) {
                    SymbolReference argument = (SymbolReference) castToExpression(aggregation.getArguments().get(0));
                    context.get().addAssignment(variable, new VariableReferenceExpression(argument.getName(), types.get(argument)));
                }
                else {
                    aggregation.getArguments().forEach(expression -> subfieldExtractor.process(castToExpression(expression), context.get()));
                }

                aggregation.getFilter().ifPresent(expression -> subfieldExtractor.process(castToExpression(expression), context.get()));

                aggregation.getOrderBy()
                        .map(OrderingScheme::getOrderByVariables)
                        .ifPresent(context.get().variables::addAll);

                aggregation.getMask().ifPresent(context.get().variables::add);
            }
```

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java [hunk 2]
```diff
@@ -527,7 +527,12 @@ private void assertPushdownSubscripts(String tableName)
     @Test
     public void testPushdownSubfields()
     {
-        assertUpdate("CREATE TABLE test_pushdown_struct_subfields(id bigint, x row(a bigint, b varchar, c double, d row(d1 bigint, d2 double)), y array(row(a bigint, b varchar, c double, d row(d1 bigint, d2 double))))");
+        assertUpdate("CREATE TABLE test_pushdown_struct_subfields(" +
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java]: Would you move this refactoring into a separate commit?

Comment (gparai) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java]: Done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java (hunk 1 above)
Comment (gparai) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java]: Thanks so much for the suggestion! We achieved the same with a lot less code. My understanding was not entirely correct and some of the code was duplicated effort of getting to what was already available in existing helper functions.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java [hunk 3]
```diff
@@ -130,8 +133,19 @@ public PlanNode visitAggregation(AggregationNode node, RewriteContext<Context> c
         {
             context.get().variables.addAll(node.getGroupingKeys());
 
-            for (AggregationNode.Aggregation aggregation : node.getAggregations().values()) {
-                aggregation.getArguments().forEach(expression -> subfieldExtractor.process(castToExpression(expression), context.get()));
+            for (Map.Entry<VariableReferenceExpression, AggregationNode.Aggregation> entry : node.getAggregations().entrySet()) {
+                VariableReferenceExpression variable = entry.getKey();
+                AggregationNode.Aggregation aggregation = entry.getValue();
+
+                // Aggregate calls with pass-through functions should allow sub-field pruning
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java]: nit: `Allow sub-field pruning to pass though arbitrary() aggregation`

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java [hunk 4]
```diff
@@ -601,6 +601,27 @@ public void testPushdownSubfields()
         assertPushdownSubfields("SELECT x.a, x.b, x.A + 2 FROM test_pushdown_struct_subfields WHERE x.B LIKE 'abc%'", "test_pushdown_struct_subfields",
                 ImmutableMap.of("x", toSubfields("x.a", "x.b")));
 
+        //Subfield pruning should pass-through arbitrary() function
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java]: - would you move the new aggregation tests into `// Aggregation` section that starts on L566?
- this query is quite complex and it is difficult to see what's exactly being tested; consider breaking it up into simpler queries that test one "use case" at a time

```
- arbitrary(x).a
- arbitrary(x.a)
- arbitrary(x).d.d1
- arbitrary(x.d).d1
- ...
```

- let's also add similar tests for arrays and maps to testPushdownArraySubscripts and testPushdownMapSubscripts respectively (this may remove the need to change the schema of the test_pushdown_struct_subfields table)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java (hunk 3 above)
Comment (gparai) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java]: Done

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java (hunk 4 above)
Comment (gparai) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java]: I have split the tests as you suggested. However, I think the combination exercises code paths which would not be exercised otherwise. Hence, there is utility in keeping these testcases. In this case, I discovered a bug in the earlier implementation where the entries were stored in a map and a column with different subfields had the old entries overwritten. I moved these tests to `testPushdownSubfieldsAssorted`. Please let me know if you feel otherwise and I can remove them.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java [hunk 5]
```diff
@@ -363,13 +363,40 @@ public void testPushdownArraySubscripts()
         assertPushdownSubfields("SELECT id, b[1] FROM test_pushdown_array_subscripts CROSS JOIN UNNEST(a) as t(unused)", "test_pushdown_array_subscripts",
                 ImmutableMap.of("b", toSubfields("b[1]")));
 
+        // Aggregation
+        assertPushdownSubfields("SELECT arbitrary(y[1]).a FROM test_pushdown_array_subscripts GROUP BY id", "test_pushdown_array_subscripts",
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java]: nit: these tests can go into `assertPushdownSubscripts` which is shared between array and map tests; this way you won't need to copy paste them

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java [hunk 6]
```diff
@@ -363,13 +363,40 @@ public void testPushdownArraySubscripts()
...
+
         // No subfield pruning
         assertPushdownSubfields("SELECT array_sort(a)[1] FROM test_pushdown_array_subscripts", "test_pushdown_array_subscripts",
                 ImmutableMap.of());
 
         assertPushdownSubfields("SELECT id FROM test_pushdown_array_subscripts CROSS JOIN UNNEST(a) as t(index) WHERE a[1] > 10 AND cardinality(b[index]) = 2", "test_pushdown_array_subscripts",
                 ImmutableMap.of());
 
+        // No pass through subfield pruning
+        assertPushdownSubfields("SELECT id, min(y[1]).a FROM test_pushdown_array_subscripts GROUP BY 1", "test_pushdown_array_subscripts",
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java]: can this go into assertPushdownSubscripts?

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java [hunk 7]
```diff
@@ -605,6 +674,13 @@ public void testPushdownSubfields()
         assertPushdownSubfields("SELECT x.a, x.b, x.A + 2 FROM test_pushdown_struct_subfields WHERE x.B LIKE 'abc%'", "test_pushdown_struct_subfields",
                 ImmutableMap.of("x", toSubfields("x.a", "x.b")));
 
+        //No pass-through subfield pruning
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java]: nit: add a space after //

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java [hunk 8]
```diff
@@ -641,6 +717,27 @@ public void testPushdownSubfieldsAssorted()
                         "a", toSubfields("a[1]"),
                         "d", toSubfields("d.d3[5]")));
 
+        //Subfield pruning should pass-through arbitrary() function
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java]: nit: add a space after //

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java [hunk 9]
```diff
@@ -641,6 +717,27 @@ public void testPushdownSubfieldsAssorted()
...
+        assertPushdownSubfields("SELECT id, " +
+                        "arbitrary(x.a), arbitrary(x).a, arbitrary(x).d.d1, arbitrary(x.d).d1, arbitrary(x.d.d2), " +
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java]: nit: perhaps, put each `arbitrary` on a separate line for readability

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java [hunk 10]
```diff
@@ -641,6 +717,27 @@ public void testPushdownSubfieldsAssorted()
...
+                        "arbitrary(z).c, " +
+                        "arbitrary(w[1][2]).e.e1, arbitrary(w[2][3].e.e2) " +
+                        "FROM test_pushdown_subfields " +
+                        "GROUP BY 1", "test_pushdown_subfields",
+                ImmutableMap.of("x", toSubfields("x.a", "x.d.d1", "x.d.d2"),
+                        "y", toSubfields("y[1].a", "y[1].d.d1", "y[2].d.d1", "y[3].d.d1"),
+                        "z", toSubfields("z.c"),
+                        "w", toSubfields("w[1][2].e.e1", "w[2][3].e.e2")));
+
+        //Subfield pruning should not pass-through other aggregate functions e.g. min() function
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java]: nit: add a space after //

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java [hunk 11]
```diff
@@ -641,6 +717,27 @@ public void testPushdownSubfieldsAssorted()
...
+        assertPushdownSubfields("SELECT id, min(x.d).d1, min(x.d.d2), min(z).c, min(z.b), min(y[1]).a, min(y[1]).d.d1, min(y[2].d.d1), min(w[1][2]).e.e1, min(w[2][3].e.e2) " +
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java]: nit: perhaps, put each `min` on a separate line for readability

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java [hunk 12]
```diff
@@ -130,8 +133,19 @@ public PlanNode visitAggregation(AggregationNode node, RewriteContext<Context> c
         {
             context.get().variables.addAll(node.getGroupingKeys());
 
-            for (AggregationNode.Aggregation aggregation : node.getAggregations().values()) {
-                aggregation.getArguments().forEach(expression -> subfieldExtractor.process(castToExpression(expression), context.get()));
+            for (Map.Entry<VariableReferenceExpression, AggregationNode.Aggregation> entry : node.getAggregations().entrySet()) {
+                VariableReferenceExpression variable = entry.getKey();
+                AggregationNode.Aggregation aggregation = entry.getValue();
+
+                //Allow sub-field pruning to pass through the arbitrary() aggregation
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java]: nit: add a space after //

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java (hunk 5 above)
Comment (gparai) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java]: Done

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java (hunk 7 above)
Comment (gparai) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java]: Done

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java (hunk 6 above)
Comment (gparai) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java]: Done

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java (hunk 8 above)
Comment (gparai) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java]: Done

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java (hunk 9 above)
Comment (gparai) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java]: Done

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java (hunk 10 above)
Comment (gparai) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java]: Done

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java (hunk 11 above)
Comment (gparai) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java]: Done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java (hunk 12 above)
Comment (gparai) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java]: Done

**Comment by gparai** [comment]
@mbasmanova thanks so much for reviewing the PR. Travis run has succeeded.

**Comment by mbasmanova** [comment]
@gparai Thank you for the contribution.
